tiledb/api/src/array/enumeration/mod.rs

```diff
@@ -0,0 +1,529 @@
+            "datatype": dtype_string,
+            "cell_val_num": cell_val_num,
+            "ordered": ordered,
+            "values": [], // TODO: Render values
+        });
+        write!(f, "{}", json)
+    }
+}
+
+impl<'c1, 'c2> PartialEq<Enumeration<'c2>> for Enumeration<'c1> {
+    fn eq(&self, other: &Enumeration<'c2>) -> bool {
+        let names_match = match (self.name(), other.name()) {
+            (Ok(mine), Ok(theirs)) => mine == theirs,
+            _ => false,
+        };
+        if !names_match {
+            return false;
+        }
+
+        let types_match = match (self.datatype(), other.datatype()) {
+            (Ok(mine), Ok(theirs)) => mine == theirs,
+            _ => false,
+        };
+        if !types_match {
+            return false;
+        }
+
+        let cell_val_num_match =
+            match (self.cell_val_num(), other.cell_val_num()) {
+                (Ok(mine), Ok(theirs)) => mine == theirs,
+                _ => false,
+            };
+        if !cell_val_num_match {
+            return false;
+        }
+
+        let ordered_match = match (self.ordered(), other.ordered()) {
+            (Ok(mine), Ok(theirs)) => mine == theirs,
+            _ => false,
+        };
+        if !ordered_match {
+            return false;
+        }
+
+        let data_match = match (self.data(), other.data()) {
+            (Ok(mine), Ok(theirs)) => mine == theirs,
+            _ => false,
+        };
+        if !data_match {
+            return false;
+        }
+
+        let offsets_match = match (self.offsets(), other.offsets()) {
+            (Ok(Some(mine)), Ok(Some(theirs))) => mine == theirs,
+            (Ok(None), Ok(None)) => true,
+            _ => false,
+        };
+        if !offsets_match {
+            return false;
+        }
+
+        true
+    }
+}
+
+pub struct Builder<'ctx> {
+    context: &'ctx Context,
+    name: String,
+    dtype: Datatype,
+    cell_val_num: u32,
+    ordered: bool,
+}
+
+impl<'ctx> Builder<'ctx> {
+    pub fn new(context: &'ctx Context, name: &str, dtype: Datatype) -> Self {
+        Builder {
+            context,
+            name: name.to_owned(),
+            dtype,
+            cell_val_num: 1,
+            ordered: false,
+        }
+    }
+
+    pub fn cell_val_num(self, cell_val_num: u32) -> Self {
+        Self {
+            cell_val_num,
+            ..self
+        }
+    }
+
+    pub fn var_sized(self) -> Self {
+        Self {
+            cell_val_num: u32::MAX,
+            ..self
+        }
+    }
+
+    pub fn ordered(self, ordered: bool) -> Self {
+        Self { ordered, ..self }
+    }
+
+    pub fn build<T>(
+        self,
+        data: &[T],
+        offsets: Option<&[u64]>,
+    ) -> TileDBResult<Enumeration<'ctx>> {
+        let mut c_enmr: *mut ffi::tiledb_enumeration_t = out_ptr!();
+        let name_bytes = self.name.as_bytes();
+        let c_name = cstring!(name_bytes);
+        let c_dtype = self.dtype.capi_enum();
```

Comment:
Should we have a little bit of code that checks `T` against `self.dtype` before proceeding?

Reply:
Not a bad idea since we already have `Datatype::is_compatible_type` available. I'll add it.

Reply:
Actually, this is kind of a bad idea. The original reason for providing this as a generic function was so that folks didn't have to do their own conversion to '&[u8]`. However, all of the enumeration strategies rely on being able to pass `'&[u8]` for any datatype. And that'll likely be the type passed whenever we provide an arrow translation API (based on my vague memory of the Array C interface).